File: solutions/c/variable-length-quantity/1/variable_length_quantity.c
#include "variable_length_quantity.h"
#include <stdint.h>
#include <stdbool.h>

int encode(const uint32_t *integers, size_t integers_len, uint8_t *output) {
    // write to `output`, return final output's length
    // `output` buffer should be enough to hold the full result

    int res_idx = 0;
    bool set_msb = false;
    int left = 0;
    int right;

    for (size_t i = 0; i < integers_len; i++) {
        uint32_t tgt = integers[i];
        left = res_idx;
        while (true) {
            int byte = tgt & 0x7F;
            tgt >>= 7;
            if (set_msb) {
                byte |= 0x80;
            }
            output[res_idx++] = byte;
            set_msb = true;
            if (tgt == 0) break;
        }
        set_msb = false;
        right = res_idx - 1;

        while (left < right) {
            uint8_t tmp = output[left];
            output[left++] = output[right];
            output[right--] = tmp;
        }
    }

    return res_idx;
}

int decode(const uint8_t *bytes, size_t buffer_len, uint32_t *output) {
    // write to `output`, return final output's length
    // return -1 if error
    // `output` buffer should be enough to hold the full result
    int res_idx = 0;
    uint32_t val = 0;

    for (size_t i = 0; i < buffer_len; i++) {
        val <<= 7;
        val |= bytes[i] & 0x7F;
        if (!(bytes[i] & 0x80)) {
            output[res_idx++] = val;
            val = 0;
        }
    }

    if ((val != 0) | (res_idx == 0)) res_idx = -1;

    return res_idx;
}

File: solutions/python/variable-length-quantity/1/variable_length_quantity.py
from math import log2


def encode(numbers):
    nums = []
    mask = 0x7F

    for n in numbers:
        if n == 0:
            nums.append(0)
            continue

        mask_offset = int(log2(n)) // 7

        while mask_offset >= 0:
            nums.append(
                (n & mask << (mask_offset * 7)) >> (mask_offset * 7)
                | (0x80 if mask_offset > 0 else 0)
            )
            mask_offset -= 1

    return nums


def decode(bytes_):
    nums = []

    to_add, incomplete = 0, False

    for n in bytes_:
        if n >> 7 != 1:
            nums.append(to_add | n)
            to_add, incomplete = 0, False
        else:
            to_add, incomplete = (to_add | (n & 0x7F)) << 7, True

    if incomplete:
        raise ValueError("incomplete sequence")

    return nums
